src/functions.php

```diff
@@ -25,15 +25,29 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
             $loop->cancelTimer($timer);
             $reject($v);
         });
+    }, function ($resolveUnused, $reject) use ($loop, &$timer, $promise) {
+        // cancelling this promise will cancel the timer and reject
+        $loop->cancelTimer($timer);
+        $reject(new \RuntimeException('Timeout cancelled'));
```

Comment:
Yes, this is different. Sorry for raising this here instead of the original issue. The idea for the current cancellation implementation has been to not give consumers the ability to reject a promise through the cancel() call but leave all responsibilities to the promise producer.

I'd vote for leaving all responsibilities to the root promise and only take them over (reject) once the timeout occurs. This means, as long as no timout occurs, everything works as we never called timeout() at all.